File: frontend/src/lib/pipeline/corpus.ts
/**
 * TF-IDF search — ported from Python corpus.py.
 * Stdlib-only implementation for finding similar reports.
 */

import { listReports, parseFrontmatter } from './storage';
import type { SearchResult } from '../types';

const WORD_RE = /[a-z][a-z0-9]+/g;
const STOPWORDS = new Set([
  'the', 'and', 'for', 'with', 'this', 'that', 'from', 'have', 'has', 'are',
  'was', 'were', 'but', 'not', 'you', 'your', 'they', 'their', 'would',
  'could', 'should', 'will', 'what', 'which', 'when', 'where', 'while',
  'into', 'than', 'then', 'them', 'also', 'any', 'all', 'can', 'may',
  'more', 'less', 'very', 'much', 'some', 'such', 'one', 'two', 'three',
  'about', 'over', 'under', 'between', 'out', 'off', 'down', 'up', 'its',
  'it', 'is', 'be', 'an', 'a', 'of', 'to', 'in', 'on', 'or', 'as', 'by',
  'at', 'if', 'we', 'us', 'our', 'i', 'do', 'does', 'did', 'so',
]);

function tokens(text: string): string[] {
  return (text.toLowerCase().match(WORD_RE) || []).filter((w) => !STOPWORDS.has(w));
}

function tfidfMatrix(docs: string[]): [Map<string, number>[], Map<string, number>] {
  const tfs: Map<string, number>[] = docs.map((doc) => {
    const counts = new Map<string, number>();
    for (const t of tokens(doc)) {
      counts.set(t, (counts.get(t) || 0) + 1);
    }
    return counts;
  });

  const df = new Map<string, number>();
  for (const tf of tfs) {
    for (const term of tf.keys()) {
      df.set(term, (df.get(term) || 0) + 1);
    }
  }

  const n = docs.length || 1;
  const idf = new Map<string, number>();
  for (const [term, cnt] of df) {
    idf.set(term, Math.log((n + 1) / (cnt + 1)) + 1);
  }

  return [tfs, idf];
}

function vec(tf: Map<string, number>, idf: Map<string, number>): Map<string, number> {
  if (tf.size === 0) return new Map();
  const norm = Math.max(...tf.values());
  const v = new Map<string, number>();
  for (const [term, count] of tf) {
    v.set(term, (count / norm) * (idf.get(term) || 0));
  }
  return v;
}

function cosine(a: Map<string, number>, b: Map<string, number>): number {
  if (a.size === 0 || b.size === 0) return 0;
  let dot = 0;
  for (const [k, va] of a) {
    if (b.has(k)) dot += va * (b.get(k) || 0);
  }
  let na = 0;
  for (const v of a.values()) na += v * v;
  let nb = 0;
  for (const v of b.values()) nb += v * v;
  na = Math.sqrt(na);
  nb = Math.sqrt(nb);
  return na && nb ? dot / (na * nb) : 0;
}

export function searchReports(query: string, top = 5): SearchResult[] {
  const reports = listReports();
  if (reports.length === 0) return [];

  const docs = reports.map((r) => r['body'] || '');
  const [tfs, idf] = tfidfMatrix([...docs, query]);
  const queryVec = vec(tfs[tfs.length - 1], idf);

  const scored: [number, Record<string, string>][] = [];
  for (let i = 0; i < tfs.length - 1; i++) {
    scored.push([cosine(vec(tfs[i], idf), queryVec), reports[i]]);
  }
  scored.sort((a, b) => b[0] - a[0]);

  return scored
    .slice(0, top)
    .filter(([sim]) => sim > 0)
    .map(([sim, r]) => ({
      slug: r['slug'] || '',
      similarity: sim,
      recommendation: (r['recommendation'] || 'explore') as 'explore' | 'build' | 'kill',
      date: r['date'] || '',
    }));
}

File: src/idea_enhancer/corpus.py
from __future__ import annotations

import math
import re
from collections import Counter
from pathlib import Path

from idea_enhancer.client import call
from idea_enhancer.storage import IDEAS_DIR, list_reports

WORD_RE = re.compile(r"[a-z][a-z0-9]+")
STOPWORDS = {
    "the", "and", "for", "with", "this", "that", "from", "have", "has", "are",
    "was", "were", "but", "not", "you", "your", "they", "their", "would",
    "could", "should", "will", "what", "which", "when", "where", "while",
    "into", "than", "then", "them", "also", "any", "all", "can", "may",
    "more", "less", "very", "much", "some", "such", "one", "two", "three",
    "about", "over", "under", "between", "out", "off", "down", "up", "its",
    "it", "is", "be", "an", "a", "of", "to", "in", "on", "or", "as", "by",
    "at", "if", "we", "us", "our", "i", "do", "does", "did", "so",
}


def _tokens(text: str) -> list[str]:
    return [w for w in WORD_RE.findall(text.lower()) if w not in STOPWORDS]


def _tfidf_matrix(docs: list[str]) -> tuple[list[Counter[str]], dict[str, float]]:
    """Returns per-doc tf vectors + idf weights. Stdlib only."""
    tfs = [Counter(_tokens(d)) for d in docs]
    df: Counter[str] = Counter()
    for tf in tfs:
        for term in tf:
            df[term] += 1
    n = len(docs) or 1
    idf = {term: math.log((n + 1) / (cnt + 1)) + 1 for term, cnt in df.items()}
    return tfs, idf


def _vec(tf: Counter[str], idf: dict[str, float]) -> dict[str, float]:
    if not tf:
        return {}
    norm = max(tf.values())
    return {term: (count / norm) * idf.get(term, 0.0) for term, count in tf.items()}


def _cosine(a: dict[str, float], b: dict[str, float]) -> float:
    if not a or not b:
        return 0.0
    common = set(a) & set(b)
    dot = sum(a[t] * b[t] for t in common)
    na = math.sqrt(sum(v * v for v in a.values()))
    nb = math.sqrt(sum(v * v for v in b.values()))
    return dot / (na * nb) if na and nb else 0.0


def _score_value(meta: dict) -> float:
    raw = meta.get("score", "")
    m = re.match(r"([0-9]+(?:\.[0-9]+)?)", str(raw))
    return float(m.group(1)) if m else -1.0


def cmd_list(root: Path = IDEAS_DIR) -> int:
    reports = list_reports(root)
    if not reports:
        print(f"No reports in {root}/. Run `enhance run \"...\"` to create one.")
        return 0
    print(f"{'date':<12} {'score':>8}  {'rec':<8}  slug")
    print("-" * 70)
    for r in sorted(reports, key=lambda r: r.get("date", ""), reverse=True):
        print(
            f"{r.get('date',''):<12} {r.get('score','-'):>8}  "
            f"{r.get('recommendation','-'):<8}  {r.get('slug','?')}"
        )
    return 0


def _find_report(slug_or_prefix: str, root: Path = IDEAS_DIR) -> dict | None:
    reports = list_reports(root)
    matches = [r for r in reports if r.get("slug", "").startswith(slug_or_prefix)]
    if len(matches) == 1:
        return matches[0]
    if len(matches) > 1:
        print(f"Ambiguous prefix {slug_or_prefix!r}; matches:")
        for m in matches:
            print(f"  {m.get('slug')}")
        return None
    print(f"No report matching {slug_or_prefix!r}.")
    return None


def cmd_show(slug: str, root: Path = IDEAS_DIR) -> int:
    r = _find_report(slug, root)
    if not r:
        return 1
    print(r["body"])
    return 0


def cmd_search(query: str, root: Path = IDEAS_DIR, top: int = 5) -> int:
    reports = list_reports(root)
    if not reports:
        print(f"No reports in {root}/.")
        return 0
    docs = [r["body"] for r in reports]
    tfs, idf = _tfidf_matrix(docs + [query])
    query_vec = _vec(tfs[-1], idf)
    scored: list[tuple[float, dict]] = []
    for tf, r in zip(tfs[:-1], reports):
        scored.append((_cosine(_vec(tf, idf), query_vec), r))
    scored.sort(key=lambda x: x[0], reverse=True)

    print(f"Query: {query!r}")
    print()
    for sim, r in scored[:top]:
        if sim <= 0:
            continue
        print(
            f"  {sim:.3f}  {r.get('date','?'):<12} "
            f"{r.get('slug','?'):<30} ({r.get('recommendation','-')})"
        )
    if not any(s > 0 for s, _ in scored):
        print("  (no matches)")
    return 0


def cmd_leaderboard(root: Path = IDEAS_DIR) -> int:
    reports = list_reports(root)
    if not reports:
        print(f"No reports in {root}/.")
        return 0
    ranked = sorted(reports, key=lambda r: _score_value(r), reverse=True)
    print(f"{'rank':>4}  {'score':>8}  {'rec':<8}  slug")
    print("-" * 70)
    for i, r in enumerate(ranked, 1):
        sv = _score_value(r)
        score_str = f"{sv:.1f}/10" if sv >= 0 else "n/a"
        print(
            f"{i:>4}  {score_str:>8}  "
            f"{r.get('recommendation','-'):<8}  {r.get('slug','?')}"
        )
    return 0


COMPARE_SYSTEM = """\
You are comparing two startup ideas head-to-head for a founder deciding \
where to allocate their next month of focus. Be specific and concrete. \
Output structure (markdown):

### Side-by-side
A short comparison covering: market size, competitive density, moat, \
why-now, and capital efficiency. One row per dimension, both ideas \
treated symmetrically.

### Where they meaningfully differ
The 2-3 dimensions where the ideas are NOT equivalent. Don't restate the \
table — explain the consequence.

### Recommendation
One paragraph: which would you pick, why, and under what condition would \
you flip. If neither is compelling vs. the founder's opportunity cost, \
say that explicitly.

Be tight. Under 500 words."""


def cmd_compare(slug_a: str, slug_b: str, root: Path = IDEAS_DIR) -> int:
    a = _find_report(slug_a, root)
    b = _find_report(slug_b, root)
    if not a or not b:
        return 1
    user = (
        f"### Idea A — {a.get('slug')}\n\n{a['body']}\n\n"
        f"### Idea B — {b.get('slug')}\n\n{b['body']}\n\n"
        "Compare per the system prompt."
    )
    print(call(system=COMPARE_SYSTEM, user=user, max_tokens=4000, effort="high"))
    return 0
